### PR TITLE
Disk cache option for gpuarray

### PR DESCRIPTION
**Pull request by abergeron**
This goes with https://github.com/Theano/libgpuarray/pull/335

also fix gh-5731
fix https://github.com/Theano/Theano/issues/4817

### CODE REVIEW

File: theano/configdefaults.py [hunk 1]
```diff
@@ -1792,7 +1792,7 @@ def default_compiledir():
     ConfigParam(
         lambda: os.path.join(config.compiledir, 'gpuarray_kernels'),
         filter=filter_base_compiledir,
-        allow_override=False)
+        allow_override=False),
```
Comment (nouiz) [inline theano/configdefaults.py]: should be documented in doc/library/config.txt

**Comment by nouiz** [comment]
The code seem good, still the doc update is missing. I'll test it later.

File: doc/library/config.txt [hunk 2]
```diff
@@ -528,6 +528,13 @@ import theano and print the config variable, as in:
     performance, check both options.
 
 
+.. attribute:: config.gpuarray.cache_path
+
+   String value
+
+   Directory to cache pre-compiled kernels for the gpuarray backend.
```
Comment (nouiz) [inline doc/library/config.txt]: Tell that it default to the subdirectory gpuarray_kernels inside theano compile dir.

Comment (abergeron) [inline doc/library/config.txt]: done

**Comment by nouiz** [comment]
When I run Theano tests, it don't speed them up and there is nothing in the
cache.

**Comment by abergeron** [comment]
I stopped refresh from deleting the gpuarray cache directory so that we can use the cache.

**Comment by slefrancois** [comment]
jenkins test this

**Comment by nouiz** [comment]
I tested it and it speed up the test of gpuarray/tests/test_basic.py when Theano cache is full compared to the master.

Ok to merge after the memory leak fix is merged and the old back-end PR.

**Comment by lamblin** [comment]
Jenkins test this again please

**Comment by nouiz** [comment]
This PR need a rebase

**Comment by lamblin** [comment]
I tried the new "Resolve conflicts" editor on GitHub, it may not have been a good idea.
I'll rebase instead if needed.

**Comment by abergeron** [comment]
The automatic change resurrected the lib.cnmem documentation so I did a manual rebase instead.

**Pull request merged**